### PR TITLE
Save  images during training

### PR DESCRIPTION
**Pull request by krisrs1128**
Saves images using weights and biases.

### CODE REVIEW

File: src/trainer.py [hunk 1]
```diff
@@ -39,11 +38,8 @@ def train(self):
       dev_loss, dev_metrics = self.evaluate(loss_f, metrics)
       train_loss, train_metrics = self.evaluate(loss_f, metrics, mode='train')
 
-      wandb.log({
-        "loss/train": train_loss,
-        "loss/dev": dev_loss
-      }, step=epoch)
-
+      print(f"epoch {epoch}/{self.config.n_epochs}\ttrain loss: {mean_loss}\tdev loss: {dev_loss}")
```
Comment (Sh-imaa) [inline src/trainer.py]: train_loss instead of mean_loss? mean_loss isn't defined, I suppose the code doesn't run now.

File: src/trainer.py [hunk 2]
```diff
@@ -67,6 +63,7 @@ def train_epoch(self, op, loss_f):
       epoch_losses.append(loss.item())
       loss.backward()
       op.step()
+      print(f"batch {i}/{len(self.train_data)}\tloss: {loss}", end="\r")
```
Comment (Sh-imaa) [inline src/trainer.py:66]: out of curiosity, any reason to use print instead of logging, I suppose logging is the best practice? 
We don't need to change it anyway, both works fine for non-production

File: src/trainer.py [hunk 3]
```diff
@@ -101,8 +99,15 @@ def evaluate(self, loss_f, metric_fs={}, mode='dev'):
             metric = fn(binary_pred, mask)
             epoch_metrics[name] += metric
 
-    return (epoch_loss / n), {name: value/n for name, value in epoch_metrics.items()}
-  
+        if self.config.store_images and i % 10  == 0:
```
Comment (Sh-imaa) [inline src/trainer.py:102]: maybe (10) should belong to the config file too, but doesn't matter now

File: src/trainer.py [hunk 4]
```diff
@@ -101,8 +99,15 @@ def evaluate(self, loss_f, metric_fs={}, mode='dev'):
...
+          img, pred, mask = img.median(axis=1).cpu(), pred.cpu(), mask.cpu()
+          for j in range(img.shape[0]):
+            merged = np.concatenate([img[j], pred[j], mask[j]], axis=1)
+            wandb_imgs.append(wandb.Image(merged))
+
+    wandb.log({f"{mode}_images": wandb_imgs})
```
Comment (Sh-imaa) [inline src/trainer.py:110]: This logs the image (with all 10/12 channels), the original mask and the logits
I think the best would be rgb (or [B5, B4, B2]), and the actual segmentation map (after Sigmoid or Softmax)

**Review by Sh-imaa — CHANGES_REQUESTED** [review]
Looks good in general, just one bug.

File: src/trainer.py (hunk 1 above)
Comment (krisrs1128) [reply src/trainer.py]: nice catch, got mixed up

File: src/trainer.py (hunk 2 above)
Comment (krisrs1128) [reply src/trainer.py:66]: Mainly because I'm counting on wandb to log the output. It will capture all the print statements in an output.log file.

File: src/trainer.py (hunk 3 above)
Comment (krisrs1128) [reply src/trainer.py:102]: Sure, we can create the issue. There are only 48 dev images, so this didn't seem like such an important parameter.

File: src/trainer.py (hunk 4 above)
Comment (krisrs1128) [reply src/trainer.py:110]: Hm, I thought I had taken care of this by taking the mean over channels in the previous two lines, but you're right, things aren't quite right. I'm looking into it.

Comment (krisrs1128) [reply src/trainer.py:110]: Yep, I just needed to apply the activations. I'm also taking three bands now, though you can't specify them in advance.

We should check to see if the activations are saturating too early in training. I'm noticing large variance, but am not sure if it's normal yet.
